From 9e2a56c7533bff2657259c4c2cae6fc08f8232ce Mon Sep 17 00:00:00 2001
From: Srujana Challa <schalla@marvell.com>
Date: Tue, 23 Mar 2021 09:04:26 +0000
Subject: [PATCH 1346/1921] crypto: cn10k: enable ME interrupts

Due to lack of ME interrupts handler in CPT PF driver, vfio-pci
driver is getting timed out with below log when it waits for
pending bit clearance during function level reset.
"vfio-pci 0002:10:00.1: timed out waiting for pending transaction;
performing function level reset anyway"
So, this patch adds support to handle ME interrupts.

Signed-off-by: Srujana Challa <schalla@marvell.com>
Change-Id: I2f31a26e4f191b8badc2063e2be538fc2bc99922
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48395
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../crypto/marvell/cn10k/cn10k_cptpf_main.c   | 94 +++++++++++++++----
 1 file changed, 78 insertions(+), 16 deletions(-)

diff --git a/drivers/crypto/marvell/cn10k/cn10k_cptpf_main.c b/drivers/crypto/marvell/cn10k/cn10k_cptpf_main.c
index 9e31fb8b38d9..78bed6f81670 100644
--- a/drivers/crypto/marvell/cn10k/cn10k_cptpf_main.c
+++ b/drivers/crypto/marvell/cn10k/cn10k_cptpf_main.c
@@ -9,34 +9,57 @@
 #define CN10K_CPT_DRV_STRING  "Marvell OcteonTX3 CPT Physical Function Driver"
 #define CN10K_CPT_DRV_VERSION "1.0"
 
-static void cptpf_enable_vf_flr_intrs(struct cn10k_cptpf_dev *cptpf)
+static void cptpf_enable_vf_flr_me_intrs(struct cn10k_cptpf_dev *cptpf)
 {
-	/* Clear interrupt if any */
+	int vfs = cptpf->max_vfs;
+
+	/* Clear FLR interrupt if any */
+	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0, RVU_PF_VFFLR_INTX(0),
+			  INTR_MASK(vfs));
+
+	/* Enable VF FLR interrupts */
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INTX(0), ~0x0ULL);
+			  RVU_PF_VFFLR_INT_ENA_W1SX(0), INTR_MASK(vfs));
+	/* Clear ME interrupt if any */
+	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0, RVU_PF_VFME_INTX(0),
+			  INTR_MASK(vfs));
+	/* Enable VF ME interrupts */
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INTX(1), ~0x0ULL);
+			  RVU_PF_VFME_INT_ENA_W1SX(0), INTR_MASK(vfs));
 
-	/* Enable VF FLR interrupts */
+	if (vfs <= 64)
+		return;
+
+	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0, RVU_PF_VFFLR_INTX(1),
+			  INTR_MASK(vfs - 64));
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INT_ENA_W1SX(0), ~0x0ULL);
+			  RVU_PF_VFFLR_INT_ENA_W1SX(1), INTR_MASK(vfs - 64));
+
+	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0, RVU_PF_VFME_INTX(1),
+			  INTR_MASK(vfs - 64));
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INT_ENA_W1SX(1), ~0x0ULL);
+			  RVU_PF_VFME_INT_ENA_W1SX(1), INTR_MASK(vfs - 64));
 }
 
-static void cptpf_disable_vf_flr_intrs(struct cn10k_cptpf_dev *cptpf)
+static void cptpf_disable_vf_flr_me_intrs(struct cn10k_cptpf_dev *cptpf)
 {
+	int vfs = cptpf->max_vfs;
+
 	/* Disable VF FLR interrupts */
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INT_ENA_W1CX(0), ~0x0ULL);
+			  RVU_PF_VFFLR_INT_ENA_W1CX(0), INTR_MASK(vfs));
+	/* Disable VF ME interrupts */
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INT_ENA_W1CX(1), ~0x0ULL);
+			  RVU_PF_VFME_INT_ENA_W1CX(0), INTR_MASK(vfs));
+
+	if (vfs <= 64)
+		return;
 
-	/* Clear interrupt if any */
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INTX(0), ~0x0ULL);
+			  RVU_PF_VFFLR_INT_ENA_W1CX(1), INTR_MASK(vfs - 64));
+
 	cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
-			  RVU_PF_VFFLR_INTX(1), ~0x0ULL);
+			  RVU_PF_VFME_INT_ENA_W1CX(1), INTR_MASK(vfs - 64));
 }
 
 static void cptpf_enable_afpf_mbox_intrs(struct cn10k_cptpf_dev *cptpf)
@@ -169,6 +192,33 @@ static irqreturn_t cptpf_vf_flr_intr(int __always_unused irq, void *arg)
 	return IRQ_HANDLED;
 }
 
+static irqreturn_t cptpf_vf_me_intr(int __always_unused irq, void *arg)
+{
+	struct cn10k_cptpf_dev *cptpf = arg;
+	int reg, vf, num_reg = 1;
+	u64 intr;
+
+	if (cptpf->max_vfs > 64)
+		num_reg = 2;
+
+	for (reg = 0; reg < num_reg; reg++) {
+		intr = cn10k_cpt_read64(cptpf->reg_base, BLKADDR_RVUM, 0,
+					RVU_PF_VFME_INTX(reg));
+		if (!intr)
+			continue;
+		for (vf = 0; vf < 64; vf++) {
+			if (!(intr & BIT_ULL(vf)))
+				continue;
+			cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
+					  RVU_PF_VFTRPENDX(reg), BIT_ULL(vf));
+			/* Clear interrupt */
+			cn10k_cpt_write64(cptpf->reg_base, BLKADDR_RVUM, 0,
+					  RVU_PF_VFME_INTX(reg), BIT_ULL(vf));
+		}
+	}
+	return IRQ_HANDLED;
+}
+
 static int cptpf_register_interrupts(struct cn10k_cptpf_dev *cptpf)
 {
 	struct pci_dev *pdev = cptpf->pdev;
@@ -197,6 +247,18 @@ static int cptpf_register_interrupts(struct cn10k_cptpf_dev *cptpf)
 	if (ret)
 		return ret;
 
+	irq = pci_irq_vector(pdev, RVU_PF_INT_VEC_VFME0);
+	ret = devm_request_irq(dev, irq, cptpf_vf_me_intr, 0,
+			       "CPTPF ME0", cptpf);
+	if (ret)
+		return ret;
+
+	irq = pci_irq_vector(pdev, RVU_PF_INT_VEC_VFME1);
+	ret = devm_request_irq(dev, irq, cptpf_vf_me_intr, 0,
+			       "CPTPF ME1", cptpf);
+	if (ret)
+		return ret;
+
 	irq = pci_irq_vector(pdev, RVU_PF_INT_VEC_AFPF_MBOX);
 	/* Register AF-PF mailbox interrupt handler */
 	ret = devm_request_irq(dev, irq, cn10k_cptpf_afpf_mbox_intr, 0,
@@ -609,7 +671,7 @@ static int cn10k_cptpf_probe(struct pci_dev *pdev,
 		goto destroy_flr;
 
 	/* Enable VF FLR interrupts */
-	cptpf_enable_vf_flr_intrs(cptpf);
+	cptpf_enable_vf_flr_me_intrs(cptpf);
 
 	/* Enable AF-PF mailbox interrupts */
 	cptpf_enable_afpf_mbox_intrs(cptpf);
@@ -641,7 +703,7 @@ static int cn10k_cptpf_probe(struct pci_dev *pdev,
 unregister_interrupts:
 	cptpf_disable_vfpf_mbox_intrs(cptpf);
 	cptpf_disable_afpf_mbox_intrs(cptpf);
-	cptpf_disable_vf_flr_intrs(cptpf);
+	cptpf_disable_vf_flr_me_intrs(cptpf);
 destroy_flr:
 	cptpf_flr_wq_destroy(cptpf);
 destroy_vfpf_mbox:
@@ -674,7 +736,7 @@ static void cn10k_cptpf_remove(struct pci_dev *pdev)
 	/* Disable AF-PF mailbox interrupt */
 	cptpf_disable_afpf_mbox_intrs(cptpf);
 	/* Disable VF FLR interrupts */
-	cptpf_disable_vf_flr_intrs(cptpf);
+	cptpf_disable_vf_flr_me_intrs(cptpf);
 	/* Unregister CPT interrupts */
 	/* Destroy AF-PF mbox */
 	cptpf_afpf_mbox_destroy(cptpf);
-- 
2.31.1

